### PR TITLE
Convert IndexedSlices to Tensors for backward ConcreteFunction calls.

### LINKED ISSUE #36236: GradientTape.gradient fails when tf.gather is used after LSTM/GRU in tf.function
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): **No**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): **Linux Stretch**
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): **binary**
- TensorFlow version (use command below): **2.1.0** and **2.0.0**
- Python version: **3.6**

**Describe the current behavior**
Consider the following simple model running a GRU, tf.gather, and a linear regression:
```python
import tensorflow as tf

inputs = tf.keras.layers.Input(shape=[None, 1], dtype=tf.float32)
hidden = tf.keras.layers.GRU(10)(inputs)
hidden = tf.gather(hidden, [0])
output = tf.keras.layers.Dense(1)(hidden)
model = tf.keras.Model(inputs=inputs, outputs=output)

@tf.function
def train(x, y):
    with tf.GradientTape() as tape:
        predictions = model(x, training=True)
        loss = tf.losses.mean_squared_error(y, predictions)
    gradients = tape.gradient(loss, model.trainable_variables)

train(tf.constant([[[1], [2], [3]]], dtype=tf.float32), tf.constant([[1]], dtype=tf.float32))
```
Both TF 2.0.0 and TF 2.1.0 fails, with a similar error message (the following is TF 2.1):
```
    ValueError: All inputs to `ConcreteFunction`s must be Tensors; on invocation of __backward_standard_gru_918, the 0-th input (IndexedSlices(indices=Tensor("Reshape_2:0", shape=(1,), dtype=int32), values=Tensor("Reshape_1:0", shape=(1, 10), dtype=float32), dense_shape=Tensor("Cast_2:0", shape=(2,), dtype=int32))) was not a Tensor.
```

**Describe the expected behavior**

The code should work.

**Other info / logs**

The problem is caused by `tf.gather` generating `tf.IndexedSlices` as a derivative. However, current LSTM/GRU are graph functions and the backprop algorithm for graph functions assumes the inputs must be `tf.Tensor`s.

One possible workaround is not to use `tf.function` and use eager. Then the code works.

Another solution is to force conversion of the derivatives from `tf.IndexedSlices` to `tf.Tensor`. The easiest solution I found is to use `* 1`, which adds operation `Mul` to the computation graph, for which the conversion from `tf.IndexedSlices` to `Tensor` works. So the following works:
```python
inputs = tf.keras.layers.Input(shape=[None, 1], dtype=tf.float32)
hidden = tf.keras.layers.GRU(10)(inputs)
hidden = tf.gather(hidden * 1, [0])
output = tf.keras.layers.Dense(1)(hidden)
model = tf.keras.Model(inputs=inputs, outputs=output)

@tf.function
def train(x, y):
    with tf.GradientTape() as tape:
        predictions = model(x, training=True)
        loss = tf.losses.mean_squared_error(y, predictions)
    gradients = tape.gradient(loss, model.trainable_variables)

train(tf.constant([[[1], [2], [3]]], dtype=tf.float32), tf.constant([[1]], dtype=tf.float32))
```

**Possible solution**
I assume an automatic conversion of `tf.IndexedSlices` to `tf.Tensor` should be performed. As a proof of concept, I changed the line https://github.com/tensorflow/tensorflow/blob/1f404dc482e3b7e4244cc1603725e555f06f8b9b/tensorflow/python/eager/function.py#L1731
to
```python
  tensor_inputs.append(ops.convert_to_tensor(arg))
```
and the code then works as expected (converting the `tf.IndexedSlices` to `tf.Tensor`). However, a proper fix is definitely more complicated.

### PR DESCRIPTION
**Pull request by foxik**
Fixes #36236.

TL;DR: When a `tf.function` is used in a computation graph, during backward pass all inputs must be Tensors. If they are IndexedSlices (which can happen when for example `tf.gather` is used), an exception is raised.

The fix automatically converts IndexedSlices to Tensors during a backward pass on a ConcreteFunction.

The conversion is currently silent, i.e., no warning is produced. I could imagine a warning being emitted.

### CODE REVIEW

**Review by alextp — APPROVED** [review]
Thank you! That was quick!
